File: Moon.Class/Component/Rotator.cs
using Fodot.Injection;
using Godot;

namespace Moon.Component;

[GlobalClass, ChildOf("CanvasItem")]
public partial class Rotator : Node
{
    [Export(PropertyHint.None, "radians_as_degrees,suffix:\u00B0/s")]
    public float Speed { get; set; } = Mathf.DegToRad(500f);
    
    [Export]
    public bool Flip { get; set; } = false;
    
    [Export]
    public bool Disabled { get; set; } = false;
    
    public enum RotatorProcessCallback { Idle, Physics }
    
    [Export]
    public RotatorProcessCallback ProcessCallback { get; set; } = RotatorProcessCallback.Physics;
}

File: Moon.Script/Rotator.fs
namespace Moon.Script

open Fodot
open Fodot.Module
open Godot
open Moon
open Moon.Component
open Moon.Utils

[<FScript(typeof<Rotator>)>]
type private RotatorScript(rot : Rotator) =
    let physics = rot.ProcessCallback = Rotator.RotatorProcessCallback.Physics
    let parent = rot |> Node.tryGetParent<CanvasItem>
    
    let update delta =
        parent
        |> Option.filter (fun _ -> rot.Disabled |> not)
        |> Option.iter (fun p ->
            let r = p |> CanvasItem.getRotation
            let d = rot.Speed * delta * (1f |> Math.flip rot.Flip)
            let r = Mathe.WrapAngle(r + d);
            p |> CanvasItem.setRotation r
        )
    
    do rot |> Engine.addDelta32Process physics update |> ignore

